util/osuser/user.go

```diff
@@ -0,0 +1,115 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+// Package osuser implements OS user lookup. It's a wrapper around os/user that
+// works on non-cgo builds.
+package osuser
+
+import (
+	"context"
+	"errors"
+	"log"
+	"os"
+	"os/exec"
+	"os/user"
+	"runtime"
+	"strings"
+	"time"
+	"unicode/utf8"
+
+	"tailscale.com/version/distro"
+)
+
+// LookupId is like os/user.LookupId but handles a few edge cases like gokrazy
+// and non-cgo lookups, and returns the user shell.
+func LookupId(uid string) (u *user.User, shell string, err error) {
+	return lookup(uid, user.LookupId)
+}
+
+// Lookup is like os/user.Lookup but handles a few edge cases like gokrazy and
+// non-cgo lookups, and returns the user shell.
+func Lookup(username string) (u *user.User, shell string, err error) {
+	return lookup(username, user.Lookup)
+}
+
+func lookup(usernameOrUID string, lookupStd func(string) (*user.User, error)) (*user.User, string, error) {
+	if runtime.GOOS != "linux" {
+		u, err := lookupStd(usernameOrUID)
+		return u, fallbackShell(), err
+	}
+
+	// No getent on Gokrazy. So hard-code the login shell.
+	if distro.Get() == distro.Gokrazy {
+		shell := "/tmp/serial-busybox/ash"
+		u, err := lookupStd(usernameOrUID)
+		if err != nil {
+			return &user.User{
+				Uid:      "0",
+				Gid:      "0",
+				Username: "root",
+				Name:     "Gokrazy",
+				HomeDir:  "/",
+			}, shell, nil
+		}
+		return u, shell, nil
+	}
+
+	// On Linux, default to using "getent" to look up users so that
+	// even with static tailscaled binaries without cgo (as we distribute),
+	// we can still look up PAM/NSS users which the standard library's
+	// os/user without cgo won't get (because of no libc hooks).
+	// But if "getent" fails, userLookupGetent falls back to the standard
+	// library anyway.
+	return userLookupGetent(usernameOrUID, lookupStd)
```

Comment:
on Linux, can we default to trying os/user first and only shell out to a child process on failure?

I guess the difference is we don't know whether the caller wants the shell info or not. Some callers do; some don't. If they do (SSH), then we have to do getent first. If they don't, it'd be better to not make child processes unnecessarily on 99% of the machines out there.

Maybe we need a separate top-level func(s) like LookupUserAndShell to declare you want the shell and only Tailscale SSH callers can use that?